### DISCUSSION TITLE
Entity relations seems to be initialized infinitely

### DISCUSSION THREAD

**Discussion by Luxiorawa** [mikro-orm/mikro-orm#3587] [Q&A] [answered]
Hello, it seems i have an issue with my entities definition.

Let's take examples of the following entities : 

Entity **Foo**
```ts
export class Foo {
  @PrimaryKey()
  public id: number;

  @OneToMany(() => FooConfig, (fooConfig) => fooConfig.foo)
  public configs = new Collection<FooConfig>(this);
}
```

Entity **FooConfig**
```ts
export class FooConfig {
  @PrimaryKey()
  public id: number;

  public key: string;

  public value: string;
  
  @ManyToOne(() => Foo)
  public foo = Foo;
}
```

In my service, when i'm requesting **Foo**, by default **FooConfig** is not populated / initialized. 

But if i want to retrieve it, i need to do the following :

```ts
const foo = await this.fooRepository.findOne(
   {
     id: fooId,
   },
   {
     populate: ["configs"],
   }
 );
```

But according to entities, there is a relation between **Foo** and **FooConfig**, but there is also a relation between **FooConfig** and **Foo**. 

Thus, the following console.log works : 

```ts
console.log("foo:", foo);
console.log("foo config :", foo.configs);
console.log("foo in config :", foo.configs[0].foo);
```

Note that this console.log doesn't compile : 

```ts
console.log("foo config in config :", foo.configs[0].foo.configs);
```

But in this logs :

```ts
console.log("foo in config :", foo.configs[0].foo);
```

I have a **FooConfig** that is initialized in foo.configs[0].foo ! So it seems that it is kinda infinitely populating / initializing my entities.

At first, i didn't understood this problem, but the problem bumped on me when i tried to implement seeding, i have a **Foo** logs that had millions and millions of lines because of this. 

Am i doing something wrong entity wise ? Or maybe there is some option to prevent this ? 

With some research, i have found the functionality [Smart Nested Populate](https://mikro-orm.io/docs/nested-populate) that is kinda close to my issue, but i'm not really sure.

**Comment by B4nan** [comment]
> Note that this console.log doesn't compile :
> `console.log("foo config in config :", foo.configs[0].foo.configs);`

With what error? I'd say it should compile.

> So it seems that it is kinda infinitely populating / initializing my entities.

Not really, in fact `foo` and `foo.configs[0].foo` are the very same object reference, so `foo === foo.configs[0].foo` is true.

https://mikro-orm.io/docs/identity-map

> i have a Foo logs that had millions and millions of lines because of this.

What do you mean by that?

**Reply by Luxiorawa** [reply]
For the error it is the following : `Property 'configs' does not exists on type 'typeof Foo'`

I see what you mean by `foo` is equal to `foo.configs[0].foo` but still, there is something i don't understand. 

I'll try to create a repo containing something i can base myself on to keep explaining.

**Reply by B4nan** [reply]
> For the error it is the following : Property 'configs' does not exists on type 'typeof Foo'

The type should be just `Foo`, not `typeof Foo`, you must have the repository wrongly typed. Lets prepare the repro so I can see what you are doing.

**Reply by Luxiorawa** [reply]
Here is the repo : https://github.com/Luxiorawa/entities-infinitely-populated

I included a docker-compose file to start the database (that is automatically launched when doing `npm run start:dev`), and a seeding file as well. 

The console.log that generate an error is in `app.service.ts` at line 31. But as you can see when logging `foo.configs[0].foo` it state that configs is initialized.

As for the initial problem, i think it will be much harder to debug / explain than i though it would be. And to answer about the "millions and millions of lines", it's not in my `Foo` entity that i have this issue (i though it was, but i retried it and i was wrong), but on more in depth entities that use this `Foo` entity as a reference.

I'll push on another branch on the same repo a more realistic example, and something that is closer to what i have, so i can explain better.

**Reply by B4nan** [reply]
Oh I just noticed you do this, thats wrong:

```ts
@ManyToOne(() => Foo)
public foo = Foo;
```

The value needs to be entity instance, not the class. 

You can do either this
```
  @ManyToOne(() => Foo)
  public foo!: Foo;
```

or this

```
  @ManyToOne({ entity: () => Foo, nullable: true })
  public foo?: Foo;
```

or this
```
  @ManyToOne(() => Foo)
  public foo = new Foo();
```

**Reply by Luxiorawa** [reply]
Oh yeah, you're right, a dumb mistake indeed, now the console.log works. But i still need to explain my initial issue aha. I just recorded the number of characters it's trying to deal with in the seeding and it's over 10 millions for something like 10 tables, and 100-200 rows inside each of them. 

But i think i know why, i'll keep going on the repo tomorrow.

**Reply by Luxiorawa** [reply]
Okay so the new branch called "seeding" is near what i currently have in my project.

I created a [database schema](https://ibb.co/fF6zS4g) so that you can better understand the whole idea.

Since i generated that schema from a database model generator, i still need to warn you that the orders table has no primary key linked to products and addresses, but store those value in JSON.

I do that because if someone order a product that has a value of let's say 10$, and that this product prices change in the futur, the order prices musn't change. So i store the value of the whole product object in JSON.

In the entity `Order` it's defined like this : 

```ts
@Entity()
export class Order {
  // Other props omited

  @Property({ type: "json" })
  public product: Product;

  @Property({ type: "json" })
  public address: Address;
}
``` 

So, if i want to do some seeding, i need to generate things from top to bottom (according to the schema). That means the following :

1) I create clients.
2) I create users, products and addresses that are linked to a client.
3) I create orders that are somehow linked to everything.

To do the link by primary key, i simply use .each() function on the Factory and then specify the client. For example 

```ts
const users = await new UserFactory(em).each((user) => {
	user.client = client
}).create(numberOfUsersPerClient)
```

Here it will create 5 users for the same client.

When i'm on the 3rd and last step, what i do is retrieving users, products and addresses randomly first. Then, i assign every relation like this :

```ts
const orders = await new OrderFactory(em).each((order) => {
	order.client = client

	const randomProductIndex = Math.floor(Math.random() * products.length)
	const randomProduct = products[randomProductIndex]
	order.product = randomProduct
	  
	const randomUserIndex = Math.floor(Math.random() * users.length)
	const randomUser = users[randomUserIndex]
	order.user = randomUser
	  
	const randomAddressIndex = Math.floor(Math.random() * addresses.length)
	const randomAddress = addresses[randomAddressIndex]
	order.address = randomAddress
}).create(numberOfOrdersPerClient)
```

There is 2 problems with doing so. The first one, is that i assign a client (which contains client.configs) to order.client, whereas it only need client, not client.configs.

The second and way more troublesome, is that since i retrieve products, users, and addresses from their factory .create() method, it returns the entity which means all relations. 

So in my products array, i have every fields of a product, but i also have the client and it's configs.

The same things occurs for address and user.

Which end up orders variable (when stringified) containing 500 000 characters, which is a lot to me. Just note that in my real project, i even end up with the seeding crashing because of memory limits, and things like that after like 10 minutes.

I tried to manually deleting each non usefull field like for example doing : 

```ts
clients.map((client) => { 
	delete client.configs
	return client
});
users.map((user) => { 
	delete user.client
	return user
});
```

But since it returns the entity, every thing i do to the entity is trying to be done on the database. So deleting client from `user.client` is trying to cut the primary key between user and client.

I could use lodash to deep clone the entity array, and then use variables containing this new array like so : 

```ts
const clientWithoutConfig = cloneDeep(clients).map((client) => {
	delete client.configs;
	return client;
})
	
const usersWithoutClient = cloneDeep(users).map((user) => {
	delete user.client;
	return user;
})
```

But isn't there a better way to achieve what i want ?

**Reply by Luxiorawa** [reply]
Sorry for the ping @B4nan but i would like to close this issue i have.

**Reply by B4nan** [reply]
So far what you described seems like correct behavior and a bunch of weird assumptions from your end. You keep talking about some 500k characters, I have no idea what that means, but whatever it means, it is not the ORM who does that, it is your code. Here you started talking about JSON properties, your reproduction does not use them at all?

Note that you should never clone entities with any 3rd party tool, that will be prone to errors.

I cant help you without understanding what your problem is. And if you keep throwing so much information at me, I will never understand it - you need to make things minimal and complete at the same time, focus on one problem at a time, and drop everything that is unrelated. I dont want to read wall of text like that, I was to see a short script I can run myself and immediately see your problem. If your issue is about JSON properties, provide a simple repro that actually uses them. Don't use any additional frameworks like Nest.js for reproductions, that only adds unnecessary complexity. It should be a single script (you can put entities and other classes to their own files, but again, in this context it only adds complexity, its better to keep everything in a single file. The file is too large? Simplify it.

edit: I see your repro is in a different branch, I dont understand why, again it only adds complexity. And again, what is in the other branch is far away from a _minimal_ repro, it is simply huge (6 entities is a lot to demonstrate any kind of problem).

**Reply by Luxiorawa** [reply]
If there is any weird assumptions from my end what about first telling me what those are so that i can clarify ?

The 500k characters i talk about is basically the variable content when stringified, what i want to point out is that the variable is too big. And i'm pretty sure it's my code that's doing it, but i also don't see any other way to make it works.

> Here you started talking about JSON properties, your reproduction does not use them at all

About my reproduction code, what do you mean by not using them ? i'm just seeding my database in the "seeding" branch, so there is nothing to really use there. There is no routes, or anything like that.

Since i'm not sure where the problem could be fixed, i just included Nest.js (which is the only way i work with mikro-orm) and seeding a database with mikro-orm. Everything else was already added in the repo for simplification (like a docker-compose script to run a database locally, where you can run the seeding to check what's happening).

The issue escalate when dealing with mutiples entities, that's why there is a bunch of them.

**Edit** : But sure i could get rid of Nest, even Docker-compose if that doesn't help you. I'll work on something

**Reply by B4nan** [reply]
> If there is any weird assumptions from my end what about first telling me what those are so that i can clarify ?

One example is the identity and propagation you dont like. Namely this:

```
the first one, is that i assign a client (which contains client.configs) to order.client, whereas it only need client, not client.configs.
```

Your entity graph needs to be complete, you need to stop thinking about "I need only that value, not the whole entity with its relations". You call this a problem, I don't see how it could be a problem. 

Ideally try to provide something as simple as one or two entities, no 3rd party deps, use as much defaults as possible. That way it will be clear what you are doing. If its not driver related, you should use in memory sqlite database so there is no docker at all. I never use the docker compose files, I have my own (the ones that the ORM uses for tests), so having something like that only adds work for me to fix the configs. But that itself is not such a huge deal, a lot of entities (and code in general) in your repro is.

Examples of simple reproductions (regression tests) are here: https://github.com/mikro-orm/mikro-orm/tree/master/tests/issues

Simple reproduction stub is also provided here: https://stackblitz.com/edit/mikro-orm-repro

> The issue escalate when dealing with mutiples entities, that's why there is a bunch of them.

The issue is in your app, and that is irrelevant, I am not here to help you (for free, in my free time) with your project. I need to see a minimal reproduction of it, that demonstrates the problem in the library.

I can understand the way you did it is simpler for you to reason about, and it is fine you went with that approach - but now you need to simplify, a lot. Start removing everything unrelated as long as your issue keeps reproducing, that way you can narrow it down.

> About my reproduction code, what do you mean by not using them ?

As mentioned in the edit - it was in another branch, so I did not see it initially.

**Reply by Luxiorawa** [reply]
@B4nan

Ok, so i came up with this : https://stackblitz.com/edit/mikro-orm-repro-iqjwef?file=index.ts

I also created a [database schema](https://i.postimg.cc/ZqhxkrWd/database-schema.png)

The idea is that one `User` can have OneToMany `Address`. 

An `Order` is done on one `User` and one `Address`. On this table, i want to have some sort of state of those entities when the `Order` was done, that's why it's in JSON and not mapped to primary keys.

I added a log at the end of the `index.ts` file so that you can better understand the problem, but the issue is that what i save in those JSON properties are the whole entities. So for a `User`, in the `Order` table i have all of users addresses. And for an `Address` i have all user data, and all it's addresses as well.

And the problem scale even more the more entities you have in the Order table, and the more entities linked you have with `User` and `Address`.

**Reply by B4nan** [reply]
Your entity definition is invalid, you can't map to-many properties to arrays, and neither they can be inversed by JSON properties. Even if you would get past the seeding issue you are trying to solve, this won't work at all, persistence will be totally broken, and everything else probably too.

**Reply by Luxiorawa** [reply]
Oh yeah sorry, forgot that it needs to be a Collection (and it is in my real code, was just a mistake there, but whatver)

I updated the link

So if i need to put those entities in a table json column, do i need to create a toJson() method on my entities and return just what i want in JSON, to then save the value of my method into the table ?

**Reply by Luxiorawa** [reply]
Okay so i made it works by just creating a toJson() function on my entities, returning an interface that only have the props i need to save in the database in JSON format. 

Just to illustrate what i did i created another repro code right here : https://stackblitz.com/edit/mikro-orm-repro-rewhvh

@B4nan seems fine or is it something i shouldn't do ? 

**EDIT** : I updated the link, i mistyped the previous one instead of the new one.

**Comment by Luxiorawa** [comment, marked as answer]
It's been a few day since the last message / comments, and the solution i found works perfectly (see [this comment](https://github.com/mikro-orm/mikro-orm/discussions/3587#discussioncomment-4017024) for reference) So i'll mark my own answer, and thus making this discussion answered.

If there is any advices or anything else, don't hesitate to ping me !

Thanks again for your help @B4nan !

EDIT : Also, sorry for the wall of text i posted, kinda new to all of this, I didn't want to frustrate you or anything else by explaining too much and things like that.
